src/Features/CSharp/Portable/IntroduceVariable/CSharpIntroduceVariableService_IntroduceLocal.cs

```diff
@@ -232,55 +227,111 @@ private Document RewriteExpressionBodiedMemberAndIntroduceLocalDeclaration(
     }
 
     private static SyntaxNode WithBlockBody(SyntaxNode node, BlockSyntax body)
+        => node switch
+        {
+            BasePropertyDeclarationSyntax baseProperty => baseProperty
+                .TryWithExpressionBody(null)
+                .WithAccessorList(AccessorList([AccessorDeclaration(SyntaxKind.GetAccessorDeclaration, body)]))
+                .TryWithSemicolonToken(Token(SyntaxKind.None)),
+            AccessorDeclarationSyntax accessor => accessor
+                .WithExpressionBody(null)
+                .WithBody(body)
+                .WithSemicolonToken(Token(SyntaxKind.None)),
+            BaseMethodDeclarationSyntax baseMethod => baseMethod
+                .WithExpressionBody(null)
+                .WithBody(body)
+                .WithSemicolonToken(Token(SyntaxKind.None)),
+            LocalFunctionStatementSyntax localFunction => localFunction
+                .WithExpressionBody(null)
+                .WithBody(body)
+                .WithSemicolonToken(Token(SyntaxKind.None)),
+            _ => throw ExceptionUtilities.UnexpectedValue(node),
+        };
+
+    private Document IntroduceLocalDeclarationIntoCompilationUnit(
+        SemanticDocument document,
+        CompilationUnitSyntax compilationUnit,
+        ExpressionSyntax expression,
+        NameSyntax newLocalName,
+        LocalDeclarationStatementSyntax declarationStatement,
+        bool allOccurrences,
+        CancellationToken cancellationToken)
     {
-        switch (node)
+        declarationStatement = declarationStatement.WithAdditionalAnnotations(Formatter.Annotation);
+
+        SyntaxNode scope = compilationUnit;
+
+        // If we're within a non-static local function, our scope for the new local declaration is expanded to include
+        // the enclosing member.
+        var localFunction = expression.GetAncestor<LocalFunctionStatementSyntax>();
+        if (localFunction is { Body: not null } && localFunction.Modifiers.Any(SyntaxKind.StaticKeyword))
+            scope = localFunction.Body;
+
+        var matches = FindMatches(
+            document, expression, document,
+            scope is ICompilationUnitSyntax
+                ? scope.ChildNodes().OfType<GlobalStatementSyntax>()
+                : [scope],
+            allOccurrences, cancellationToken);
+        Debug.Assert(matches.Contains(expression));
+
+        var firstAffectedExpression = matches.OrderBy(m => m.SpanStart).First();
+
+        var editor = new SyntaxEditor(compilationUnit, document.Project.Solution.Services);
+
+        // Parenthesize the variable, and go and replace anything we find with it. NOTE: we do not want elastic trivia
+        // as we want to just replace the existing code as is, while preserving the trivia there.  We do not want to
+        // update it.
+        var replacement = editor.Generator.AddParentheses(newLocalName, includeElasticTrivia: false);
+        foreach (var match in matches)
```

Comment:
>matches [](http://example.com/codeflow?start=30&length=7)

no reverse ordering needed?

Reply:
no.  tehre can be no nesting with "intro variable for all occurrences"

Reply:
definitionally, if there was a node wrapping another node, it will represent some larger tree.